keps/sig-autoscaling/2021-scale-from-zero/README.md

```diff
@@ -0,0 +1,934 @@
+
+Feature gates are typically disabled by setting the flag to `false` and
+restarting the component. No other changes should be necessary to disable the
+feature.
+
+NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
+-->
+
+
+Yes. To downgrade the cluster to version that does not support scale-to-zero feature:
+
+1. Make sure there are no hpa objects with minReplicas=0 and maxReplicas=0. Here is a oneliner to update it to 1:
+
+    `$ kubectl get hpa --all-namespaces  --no-headers=true | awk  '{if($6==0) printf "kubectl patch hpa/%s --namespace=%s -p \"{\\\"spec\\\":{\\\"minReplicas\\\":1,\\\"maxReplicas\\\":1}}\"\n", $2, $1 }' | sh`
+2. Disable `HPAScaleToZero` feature gate
+3. In case step 1. has been omitted, workloads might be stuck with `replicas: 0` and need to be manually scaled up to `replicas: 1` to re-enable autoscaling.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+Nothing, the feature can be re-enabled without problems.
+
+###### Are there any tests for feature enablement/disablement?
+
+<!--
+The e2e framework does not currently support enabling or disabling feature
+gates. However, unit tests in each component dealing with managing data, created
+with and without the feature, are necessary. At the very least, think about
+conversion tests if API types are being modified.
+-->
+
+There currently unit tests for the alpha cases and tests planned to be added for the new functionality.
+
+### Rollout, Upgrade and Rollback Planning
+
+<!--
+This section must be completed when targeting beta to a release.
+-->
+
+As this is a new field every usage is opt-in. In case the kubernetes version is downgraded, currently scaled to 0 workloads might need to be manually scaled to 1 as the controller would treat them as
+paused otherwise.
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+<!--
+Try to be as paranoid as possible - e.g., what if some components will restart
+mid-rollout?
+
+Be sure to consider highly-available clusters, where, for example,
+feature flags will be enabled on some API servers and not others during the
+rollout. Similarly, consider large clusters and how enablement/disablement
+will rollout across nodes.
+-->
+
+There are no expected side-effects when the rollout fails as new `enableScaleToZero` flag should only be enabled once the version upgraded completed and should be disabled before attempting a rollback.
+
+In case this is missed, HPA for deployments with zero replicas will be disabled until replicas have been raised explicitly to at least `1`.
+
+###### What specific metrics should inform a rollback?
+
+<!--
+What signals should users be paying attention to when the feature is young
+that might indicate a serious problem?
+-->
+
+If workloads aren't scaled up from 0 despite the scaling condition being meet, an operator should rollback this feature and manually scale an affected workload back to `1`.
```

Comment:
Is there a metric or something? Imagine you run 50,000 clusters and your users may or may not be using this feature you enabled. How can you easily tell if it's working or needs to be turned off? Typically this is a metric.

Reply:
Working autoscaling can generally be verified by comparing [kube_horizontalpodautoscaler_status_current_replicas](https://github.com/kubernetes/kube-state-metrics/blob/main/internal/store/horizontalpodautoscaler.go#L303C4-L303C56) and [kube_horizontalpodautoscaler_status_desired_replicas](https://github.com/kubernetes/kube-state-metrics/blob/main/internal/store/horizontalpodautoscaler.go#L391C4-L391C49) and some alerting when those values aren't matching within a timeframe. Scale to zero could be observed in a similar fashion.